nixos/etesync-dav: init module

matt-snider opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

[etesync-dav](https://github.com/etesync/etesync-dav) was recently packaged in #76886 and since then I've been using this module locally. I think it could be useful to others.

One thing to consider is that the current implementation requires that the user go to the management panel once the service is running (default `http://localhost:37358`), use their etesync credentials to login, and then copy the generated password for use in other programs that want to use etesync-dav to bridge the connection to etesync. It might be possible to have the credentials generated on startup from some additional module options (e.g. `etesyncUsername`, `etesyncPasswordCommand`) similarly to [here](https://github.com/etesync/etesync-dav/tree/master/examples/systemd-sandbox#how-should-i-generate-the-credential-files-and-radicale-configuration), but I'm not sure this would work nicely, and the generated password still needs to be passed back to the user securely.  
 
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

matt-snider committed:
nixos/etesync-dav: init module

aanderse commented on nixos/modules/services/misc/etesync-dav.nix:
```diff
@@ -0,0 +1,69 @@
+    options.services.etesync-dav = {
+      enable = mkEnableOption "etesync-dav";
+
+      package = mkOption {
```
It looks like we only have 1 variant of this package. I think this option might not provide that much value.

matt-snider replied:
I've been adding it to all modules I contribute because I figured it was a best practice. But good point, I will remove it.

aanderse commented on nixos/modules/services/misc/etesync-dav.nix:
```diff
@@ -0,0 +1,69 @@
+          ETESYNC_URL = cfg.apiUrl;
+          ETESYNC_DATA_DIR = "/var/lib/etesync-dav";
+        };
+
```
There *might* be one or two more options you could from https://github.com/etesync/etesync-dav/blob/master/examples/systemd-sandbox/etesync-dav%40.service but I haven't verified that and could easily be wrong.

matt-snider replied:
I originally based that list on the [config.py](https://github.com/etesync/etesync-dav/blob/a90a1760fda7f38c8e432503adbc35c9335749b2/etesync_dav/config.py) file. 

`ETESYNC_CONFIG_DIR` is a legacy option and no longer necessary, and several of the other options could result in the files being outside of the `StateDirectory` which would also cause errors (e.g. `ETESYNC_DATABASE_FILE`, `CREDS_FILE`, etc).

As you mentioned in a previous comment, allowing SSL support would be good, so I think `SSL_CERT_FILE` and `SSL_KEY_FILE` should be added. I'll update the PR.

aanderse replied:
I was actually referring to `systemd` options, but ssl options sound great too :+1:

aanderse replied:
~~I think you need to add `SSL_CERT_FILE` and `SSL_KEY_FILE` to `environment`.~~ Scratch that. After looking at [config.py](https://github.com/etesync/etesync-dav/blob/a90a1760fda7f38c8e432503adbc35c9335749b2/etesync_dav/config.py) I notice the default value is set to that.

aanderse commented on nixos/modules/services/misc/etesync-dav.nix:
```diff
@@ -0,0 +1,69 @@
+      host = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = "etesync-dav host";
```
Please be consistent with use of `.` in `description` field.

aanderse reviewed (commented):
What a great module :+1: Very clean and simple! A few points:
- regarding the steps required I think that sounds fine, but is there a need to document/describe to the user what to do, or is that obvious for users of this software?
- I'm not familiar with this software but it sounds like something that should be behind ssl if exposed outside of `localhost`. Is this correct?

matt-snider commented:
Hi Aaron, thanks for the review!

> regarding the steps required I think that sounds fine, but is there a need to document/describe to the user what to do, or is that obvious for users of this software?

I think that should be obvious from the [README](https://github.com/etesync/etesync-dav/#configuration-and-running) of the software itself. I'm still new to the Nix Community though so correct me if I'm wrong.

> I'm not familiar with this software but it sounds like something that should be behind ssl if exposed outside of `localhost`. Is this correct?

As noted above, I'll add support for this via `SSL_CERT_FILE` and `SSL_KEY_FILE`

matt-snider commented:
@aanderse I updated the PR. The SSL cert handling was modeled after [the syncthing module](https://github.com/NixOS/nixpkgs/blob/d04396da8bf320f41a2bd939f2f4eb83a586b3b2/nixos/modules/services/networking/syncthing.nix#L470). 

Let me know what you think!

aanderse commented on nixos/modules/services/misc/etesync-dav.nix:
```diff
@@ -0,0 +1,94 @@
+      };
+
+      port = mkOption {
+        type = types.int;
```
`types.port`

matt-snider replied:
Good catch!

aanderse reviewed (commented):
@matt-snider is there an opportunity to write a test for this module? That would really make this PR a home run :baseball:

matt-snider commented:
@aanderse Great idea on the test! I found [a few issues](https://github.com/NixOS/nixpkgs/compare/e011cee392ba371cbea6ff2f3c2c8ca90c1fab08..3dcb2b8d78eec8aa6e85e86251883ce9ee695f9d) as a result:

* I forgot to add it to the module-list
* In my manual tests the certs happened to be named properly (etesync.key, etesync.crt). The `install` command should actually enforce this though. 

Anyways, the test cases are quite simple and basically just check that the login screen is reachable. I don't know if there is much else that can be tested. 

Note: I commented out a test that checks for the redirect from http -> https. I noticed this doesn't currently work. There is however already [an unreleased fix](https://github.com/etesync/etesync-dav/commit/7ecda8894273f3d0419edbb93f819837aa04ace5), so it _should_ work in the next release. @Valodim: Just an FYI about the test since you are the maintainer of the etesync-dav package. It can likely be added in the next release (> 0.16.0)

matt-snider commented:
Hey @aanderse can you take a look again when you have some time?

aanderse commented:
Sorry @matt-snider this dropped off my radar.
@GrahamcOfBorg test etesync-dav

matt-snider commented:
@aanderse No problem. 

Hmm tests were passing before, but it looks like etesync-dav was upgraded to 0.16.0 in the meantime (see #85061). I'll have a look and see if I can also uncomment the other test for the https redirect.

matt-snider commented:
@aanderse  I rebased and fixed the test.

The problem was that I referenced the `etesync-dav` user and group to copy the keys in `ExecStartPre` and on the very first run, the user & group weren't created yet (I guess due to `DynamicUser=true`). I'm not sure why it was working before, but it might have been due to running with `-A driver` and accidentally reusing existing VM state.

Anyways, I realized I don't need to do this step as root (via `ExecStartPre=+...`) or use `install`. Since the script runs as the correct user anyways `cp` looks like it works just fine.

aanderse commented:
@GrahamcOfBorg test etesync-dav
@Valodim are you interested in leaving a review?

Valodim commented:
Oh wow, this one kind of slipped past me! At a first glance this looks great! I'll try to give it a full review soon.

Meanwhile I updated etesync-dav #93556, which broke somewhere along the way. Quick review on that would be welcome to bring this back on track.

matt-snider commented:
@Valodim A review would be great - thanks! Since you've updated to 0.20, the other test for the http to https redirect should also work now. I'lll try it out soon

Valodim commented:
So I started reviewing this, and as noted before things seem mostly straightforward :)

However, I wondered one thing: Is setup for non-local access really a use case we want to support in this module? etesync-dav is intended to be run as a local proxy, so that the dav client becomes the actual cryptographic endpoint, to make things "end to end".
There are a bunch of options in this module (`openFirewall`, `sslCertificate`, `sslCertificateKey`, `host`) that don't fit this idea. Do either of you use etesync-dav in this way, personally?

As a concrete proposal for discussion: We could drop those options listed above in favor of a generic `settings` one, which is passed straight into the environment. That gives flexibility to folks with any use case, but also avoids having more than half the options of this module in the "not really the intended way of using it" bucket.

For comparison, I pushed the etesync-dav module I use in home-manager for the same purpose: https://github.com/rycee/home-manager/pull/1398

aanderse commented:
ping @matt-snider

matt-snider commented:
@Valodim Thanks for having a look and making some suggestions!

> However, I wondered one thing: Is setup for non-local access really a use case we want to support in this module? etesync-dav is intended to be run as a local proxy, so that the dav client becomes the actual cryptographic endpoint, to make things "end to end".
There are a bunch of options in this module (openFirewall, sslCertificate, sslCertificateKey, host) that don't fit this idea. Do either of you use etesync-dav in this way, personally?

That's a good point - it's something I hadn't considered. I certainly don't use etesync-dav this way, but I'm not sure we should lump all of those options into the "non-local access" bucket. SSL is properly supported by etesync-dav and it seems like it might make sense for us to support it too. [In the windows instructions](https://github.com/etesync/etesync-dav/blob/master/win10-instructions.md#setup-ssl) it's mentioned that some clients won't work without SSL configured. Although Windows-specific instructions are irrelevant to us, I could foresee other clients having the same quirk.

> As a concrete proposal for discussion: We could drop those options listed above in favor of a generic settings one, which is passed straight into the environment. That gives flexibility to folks with any use case, but also avoids having more than half the options of this module in the "not really the intended way of using it" bucket.

If we decide to keep the SSL options, then it's really just `host` and `openFirewall` that could potentially be eliminated. If non-local use is, as you say "not really the intended way of using it", then I think `openFirewall` can be dropped. It's also easy for users to work around if they do end up needing it. I'm not sure which local-only use cases would require changing `host`, but I'm not certain that there aren't any, and it's a one-liner that sets an env var, so supporting it is easy. My inclination would be to leave it.

What do you think?

aanderse commented:
Oh wow... this sure dropped off the radar. @Valodim are you able to provide the feedback requested by @matt-snider?

matt-snider commented:
@aanderse I don't know that we will get a response from @Valodim -- do you have an opinion on the above? The main part is this:

> If we decide to keep the SSL options, then it's really just host and openFirewall that could potentially be eliminated. If non-local use is, as you say "not really the intended way of using it", then I think openFirewall can be dropped. It's also easy for users to work around if they do end up needing it. I'm not sure which local-only use cases would require changing host, but I'm not certain that there aren't any, and it's a one-liner that sets an env var, so supporting it is easy. My inclination would be to leave it.

Basically, we are just mapping nix options to ETESYNC environment variables for the most part, and I'm inclined to allow setting all of the supported variables. `openFirewall` is something I could drop - I don't really have an opinion on that since I don't use it.

matt-snider commented:
@aanderse Another question: on another PR is submitted, there was a [comment](https://github.com/NixOS/nixpkgs/pull/82480#discussion_r544619689) about how the tests aren't very useful. The same could probably be said about this one. I still didn't receive an answer on that comment about alternative ways to test. Do you think I should remove or change the tests for this PR too?

aanderse commented:
@matt-snider I'll trust your judgement on this one - either way you see fit. Aside from the merge conflict I think this is ready to merge :+1:

*Possible* **future** PRs against this module include:
- add `systemd` hardening options from the [upstream unit](https://github.com/etesync/etesync-dav/blob/9ebfdd588f36d5eb3a49e9397ed5a192bcb07cc7/examples/systemd-sandbox/etesync-dav%40.service#L43-L63)
- utilize [`LoadCredential`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#LoadCredential=ID:PATH) to avoid the need for an `ExecStartPre`, as well as having key files with secure permissions

Thanks for your work (and patience!) on this.

matt-snider commented:
@aanderse I will remove the tests, as I think the criticism in the above linked comment holds about them testing relatively little compared to the cost of evaluation.

Those are some interesting suggestions and something I will think about adding going forwards.

In particular, I wasn't aware of the systemd `LoadCredential` directive - this is very cool! My initial reaction was to ask whether there are any plans to use this for NixOS secret management, but then I searched and came across [RFC 59](https://github.com/NixOS/rfcs/pull/59#issuecomment-624723733). Thank you for sharing that!

I will look into updating the PR ASAP!

matt-snider commented:
@aanderse I removed the test so there are no conflicts anymore. I think we can merge it

aanderse commented:
@matt-snider I'm really sorry I missed this before another merge conflict came up. I moved last week so I wasn't able to react as fast as I would have liked. :disappointed:

matt-snider commented:
@aanderse Hey thanks for staying on this. Now I'm the one who's sorry for the slow response! I rebased and fixed the conflict

Congrats on the move :-)

aanderse commented:
Thanks @matt-snider!

matt-snider commented:
Thank you for all your work reviewing this @aanderse !
